File: sanity_studio/schemaTypes/mark_1_schemas/about.ts

export default {
  name: 'about',
  title: 'About ME',
  type: 'document',
  fields: [
    {
      name: "firstName",
      type: "string",
      title: "First Name",
      validation: (Rule) => Rule.required().error("First Name is mandatory."),
    },
    {
      name: "secondName",
      type: "string",
      title: "Second Name",
    },
    {
      name: "lastName",
      type: "string",
      title: "Last Name",
    },
    {
      name: "age",
      type: "number",
      title: "Age",
      validation: (Rule) => Rule.min(0).error("Age cannot be negative."),
    },
    {
      name: "dateOfBirth",
      type: "date",
      title: "Date of Birth",
      options: {
        dateFormat: "YYYY-MM-DD",
        calendarTodayLabel: "Today",
      },
    },
    {
      name: "gender",
      type: "string",
      title: "Gender",
      options: {
        list: [
          { title: "Male", value: "male" },
          { title: "Female", value: "female" },
          { title: "Other", value: "other" },
        ],
      },
    },
    {
      name: "nationality",
      type: "string",
      title: "Nationality",
    },
    {
      name: "detailDescription",
      type: "text",
      title: "Detail Description",
      description: "Enter a detailed description.",
    },
    {
      name: "contact",
      type: "array",
      title: "Contact Information",
      of: [
        {
          type: "object",
          title: "Contact",
          fields: [
            {
              name: "contactName",
              type: "string",
              title: "Contact Name",
              validation: (Rule) =>
                Rule.required().error("Contact Name is mandatory."),
            },
            {
              name: "contactDetails",
              type: "string",
              title: "Contact Details",
            },
            {
              name: "contactIcon",
              type: "image",
              title: "Contact Icon",
              options: {
                hotspot: true, // Enables image cropping in the studio
              },
              description: "Upload an icon for this contact method.",
            },
            {
              name: "hasLink",
              type: "boolean",
              title: "Has Link",
            },
            {
              name: "contactLink",
              type: "url",
              title: "Contact Link",
              hidden: ({ parent }) => !parent?.hasLink, // Show only if hasLink is true
            },
          ],
        },
      ],
    },
    {
      name: "shortDescription",
      type: "text",
      title: "Short Description",
      description: "A brief 2-3 line description.",
      validation: (Rule) =>
        Rule.max(300).warning("Short Description should be concise."),
    },
    {
      name: "coverImage",
      type: "image",
      title: "Cover Image",
      options: {
        hotspot: true, // Enables image cropping in the studio
      },
      description: "Upload a cover image for the person.",
    },
    {
      name: "avatarImage",
      type: "image",
      title: "Avatar Image",
      options: {
        hotspot: true, // Enables image cropping in the studio
      },
      description: "Upload an avatar image for the person.",
    },
  ],
};

File: app/lib/api/models/about.dart
import 'package:json_annotation/json_annotation.dart';

part 'about.g.dart';

@JsonSerializable(explicitToJson: true)
class About {
  @JsonKey(name: '_createdAt')
  final String createdAt;

  @JsonKey(name: '_id')
  final String id;

  @JsonKey(name: '_rev')
  final String rev;

  @JsonKey(name: '_type')
  final String type;

  @JsonKey(name: '_updatedAt')
  final String updatedAt;

  final int age;
  final List<Contact> contact;
  final ImageAsset coverImage;
  final String dateOfBirth;
  final String detailDescription;
  final String firstName;
  final String gender;
  final String lastName;
  final String nationality;
  final String secondName;
  final String shortDescription;

  About({
    required this.createdAt,
    required this.id,
    required this.rev,
    required this.type,
    required this.updatedAt,
    required this.age,
    required this.contact,
    required this.coverImage,
    required this.dateOfBirth,
    required this.detailDescription,
    required this.firstName,
    required this.gender,
    required this.lastName,
    required this.nationality,
    required this.secondName,
    required this.shortDescription,
  });

  factory About.fromJson(Map<String, dynamic> json) => _$AboutFromJson(json);
  Map<String, dynamic> toJson() => _$AboutToJson(this);
}

@JsonSerializable()
class Contact {
  @JsonKey(name: '_key')
  final String key;
  final String? contactDetails;
  final ImageAsset contactIcon;
  final String contactName;
  final bool hasLink;
  final String? contactLink;

  Contact({
    required this.key,
    this.contactDetails,
    required this.contactIcon,
    required this.contactName,
    required this.hasLink,
    this.contactLink,
  });

  factory Contact.fromJson(Map<String, dynamic> json) =>
      _$ContactFromJson(json);
  Map<String, dynamic> toJson() => _$ContactToJson(this);
}

@JsonSerializable()
class ImageAsset {
  @JsonKey(name: '_type')
  final String type;
  final Asset asset;

  ImageAsset({required this.type, required this.asset});

  factory ImageAsset.fromJson(Map<String, dynamic> json) =>
      _$ImageAssetFromJson(json);
  Map<String, dynamic> toJson() => _$ImageAssetToJson(this);
}

@JsonSerializable()
class Asset {
  @JsonKey(name: '_ref')
  final String ref;
  @JsonKey(name: '_type')
  final String type;

  Asset({required this.ref, required this.type});

  factory Asset.fromJson(Map<String, dynamic> json) => _$AssetFromJson(json);
  Map<String, dynamic> toJson() => _$AssetToJson(this);
}
